api/pkg/s3/datatype/httprange.go

```diff
@@ -0,0 +1,141 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package datatype
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+const (
+	byteRangePrefix = "bytes="
+)
+
+// Valid byte position regexp
+var validBytePos = regexp.MustCompile(`^[0-9]+$`)
+
+// ErrorInvalidRange - returned when given range value is not valid.
+var ErrorInvalidRange = errors.New("Invalid range")
+
+// HttpRange specifies the byte range to be sent to the client.
+type HttpRange struct {
+	OffsetBegin  int64
+	OffsetEnd    int64
+	ResourceSize int64
+}
+
+// String populate range stringer interface
+func (hrange HttpRange) String() string {
+	return fmt.Sprintf("bytes %d-%d/%d", hrange.OffsetBegin, hrange.OffsetEnd, hrange.ResourceSize)
+}
+
+// getlength - get length from the range.
+func (hrange HttpRange) GetLength() int64 {
+	return 1 + hrange.OffsetEnd - hrange.OffsetBegin
+}
+
+func ParseRequestRange(rangeString string, resourceSize int64) (hrange *HttpRange, err error) {
+	// TODO handle multi-range request
+	// see https://tools.ietf.org/html/rfc7233
+
+	// Return error if given range string doesn't start with byte range prefix.
+	if !strings.HasPrefix(rangeString, byteRangePrefix) {
+		return nil, fmt.Errorf("'%s' does not start with '%s'", rangeString, byteRangePrefix)
+	}
+
+	// Trim byte range prefix.
+	byteRangeString := strings.TrimPrefix(rangeString, byteRangePrefix)
+
+	// Check if range string contains delimiter '-', else return error. eg. "bytes=8"
+	sepIndex := strings.Index(byteRangeString, "-")
+	if sepIndex == -1 {
+		return nil, fmt.Errorf("'%s' does not have a valid range value", rangeString)
+	}
+
+	offsetBeginString := byteRangeString[:sepIndex]
+	offsetBegin := int64(-1)
+	// Convert offsetBeginString only if its not empty.
+	if len(offsetBeginString) > 0 {
+		if !validBytePos.MatchString(offsetBeginString) {
+			return nil, fmt.Errorf("'%s' does not have a valid first byte position value", rangeString)
+		}
+
+		if offsetBegin, err = strconv.ParseInt(offsetBeginString, 10, 64); err != nil {
+			return nil, fmt.Errorf("'%s' does not have a valid first byte position value", rangeString)
+		}
+	}
+
+	offsetEndString := byteRangeString[sepIndex+1:]
+	offsetEnd := int64(-1)
+	// Convert offsetEndString only if its not empty.
+	if len(offsetEndString) > 0 {
+		if !validBytePos.MatchString(offsetEndString) {
+			return nil, fmt.Errorf("'%s' does not have a valid last byte position value", rangeString)
+		}
+
+		if offsetEnd, err = strconv.ParseInt(offsetEndString, 10, 64); err != nil {
+			return nil, fmt.Errorf("'%s' does not have a valid last byte position value", rangeString)
+		}
+	}
```

Comment:
duplicate code...please refactor into a method.
used for offsetBeginString and offsetEndString

Reply:
@rhsakarpos , Thanks, I will refactor it